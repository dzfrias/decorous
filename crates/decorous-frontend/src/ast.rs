use std::{borrow::Cow, fmt, path::Path};

use heck::ToSnekCase;
use itertools::Itertools;
use rslint_parser::SyntaxNode;

use crate::{css::ast::Css, location::Location};

/// The abstract syntax tree representation of decorous source code.
///
/// This struct has three main components (all with respective getter methods):
/// 1. [`nodes`](Self::nodes), for holding HTML markup.
/// 2. [`script`](Self::script), for holding JavaScript.
/// 3. [`css`](Self::css), for holding CSS.
///
/// One of the best uses of `DecorousAst` is to hold data, read some parts, and eventually use
/// [`into_components()`](DecorousAst::into_components()) when the time has come to apply more
/// complex transformations that require ownership.
#[derive(Debug)]
pub struct DecorousAst<'a> {
    pub nodes: Vec<Node<'a, Location>>,
    pub script: Option<SyntaxNode>,
    pub css: Option<Css>,
    pub wasm: Option<Code<'a>>,
    pub comptime: Option<Code<'a>>,
}

/// A node of the [AST](DecorousAst).
///
/// It contains [metadata](`Self::metadata()`) (of type `T`), and the
/// actual node data, retrieved by [`node_type()`](`Self::node_type()`).
#[derive(Debug, Clone, PartialEq)]
pub struct Node<'a, T> {
    pub metadata: T,
    pub node_type: NodeType<'a, T>,
}

#[derive(Debug, Clone, PartialEq)]
pub enum NodeType<'a, T> {
    Element(Element<'a, T>),
    Text(Text<'a>),
    Comment(Comment<'a>),
    SpecialBlock(SpecialBlock<'a, T>),
    Mustache(Mustache),
}

#[derive(Debug, Clone, PartialEq, Hash)]
pub struct Mustache(pub SyntaxNode);

impl fmt::Display for Mustache {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl std::ops::Deref for Mustache {
    type Target = SyntaxNode;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

#[derive(Debug, Clone, PartialEq, Hash, Copy)]
pub struct Text<'a>(pub &'a str);

impl fmt::Display for Text<'_> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl<'a> std::ops::Deref for Text<'a> {
    type Target = &'a str;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

#[derive(Debug, Clone, PartialEq, Hash, Copy)]
pub struct Comment<'a>(pub &'a str);

impl<'a> std::ops::Deref for Comment<'a> {
    type Target = &'a str;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct Element<'a, T> {
    pub tag: &'a str,
    pub attrs: Vec<Attribute<'a>>,
    pub children: Vec<Node<'a, T>>,
}

#[derive(Debug, Clone, PartialEq)]
pub enum SpecialBlock<'a, T> {
    For(ForBlock<'a, T>),
    If(IfBlock<'a, T>),
    Use(UseBlock<'a>),
}

#[derive(Debug, Clone, PartialEq)]
pub struct ForBlock<'a, T> {
    pub binding: &'a str,
    pub index: Option<&'a str>,
    pub expr: SyntaxNode,
    pub inner: Vec<Node<'a, T>>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct IfBlock<'a, T> {
    pub expr: SyntaxNode,
    pub inner: Vec<Node<'a, T>>,
    pub else_block: Option<Vec<Node<'a, T>>>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct UseBlock<'a> {
    pub path: &'a Path,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Attribute<'a> {
    EventHandler(EventHandler<'a>),
    Binding(&'a str),
    KeyValue(&'a str, Option<AttributeValue<'a>>),
}

#[derive(Debug, Clone, PartialEq)]
pub struct EventHandler<'a> {
    pub event: &'a str,
    pub expr: SyntaxNode,
}

#[derive(Debug, Clone, PartialEq)]
pub enum AttributeValue<'a> {
    Literal(Cow<'a, str>),
    JavaScript(SyntaxNode),
}

#[derive(Debug, Clone, PartialEq)]
pub enum CollapsedChildrenType<'a> {
    Text(&'a str),
    Html(String),
}

#[derive(Debug)]
pub struct Code<'a> {
    pub lang: &'a str,
    pub body: &'a str,
    pub offset: usize,
    pub comptime: bool,
}
impl<'a, T> Element<'a, T> {
    pub fn descendents(&'a self) -> NodeIter<'a, T> {
        NodeIter::new(&self.children)
    }

    pub fn descendents_mut<F>(&'a mut self, f: &mut F)
    where
        F: FnMut(&mut Node<'a, T>),
    {
        for child in &mut self.children {
            f(child);
            if let NodeType::Element(elem) = &mut child.node_type {
                elem.descendents_mut(f);
            }
        }
    }

    pub fn has_immediate_mustache(&self) -> bool {
        self.children
            .iter()
            .any(|child| matches!(child.node_type, NodeType::Mustache(_)))
    }

    pub fn js_valid_tag_name(&self) -> Cow<'a, str> {
        if self.tag.contains('-') {
            Cow::Owned(self.tag.to_snek_case())
        } else {
            Cow::Borrowed(self.tag)
        }
    }
}

impl<'a, T> Node<'a, T> {
    pub fn new(ty: NodeType<'a, T>, metadata: T) -> Self {
        Self {
            metadata,
            node_type: ty,
        }
    }

    /// Recursively cast each the metadata field of each node into a new type. The provided
    /// function receives the previous metadata of the node, and should return the corresponding new
    /// metadata.
    pub fn cast_meta<U, F>(self, transfer_func: &mut F) -> Node<'a, U>
    where
        F: FnMut(&Node<'a, T>) -> U,
    {
        macro_rules! cast_children {
            ($children_vec:expr, $transfer:expr) => {
                $children_vec
                    .into_iter()
                    .map(|node| node.cast_meta($transfer))
                    .collect()
            };
        }

        let new_meta = transfer_func(&self);

        // A bit verbose, but it's the only way to do a full type cast throughout the entire AST
        match self.node_type {
            NodeType::SpecialBlock(block) => Node {
                metadata: new_meta,
                node_type: NodeType::SpecialBlock(match block {
                    SpecialBlock::If(if_block) => SpecialBlock::If(IfBlock {
                        expr: if_block.expr,
                        inner: cast_children!(if_block.inner, transfer_func),
                        else_block: if_block
                            .else_block
                            .map(|nodes| cast_children!(nodes, transfer_func)),
                    }),
                    SpecialBlock::For(for_block) => SpecialBlock::For(ForBlock {
                        inner: cast_children!(for_block.inner, transfer_func),
                        binding: for_block.binding,
                        index: for_block.index,
                        expr: for_block.expr,
                    }),
                    SpecialBlock::Use(use_block) => SpecialBlock::Use(use_block),
                }),
            },
            NodeType::Element(elem) => Node {
                metadata: new_meta,
                node_type: NodeType::Element(Element {
                    tag: elem.tag,
                    attrs: elem.attrs,
                    children: cast_children!(elem.children, transfer_func),
                }),
            },
            NodeType::Text(text) => Node {
                metadata: new_meta,
                node_type: NodeType::Text(text),
            },
            NodeType::Comment(comment) => Node {
                metadata: new_meta,
                node_type: NodeType::Comment(comment),
            },
            NodeType::Mustache(syntax_node) => Node {
                metadata: new_meta,
                node_type: NodeType::Mustache(syntax_node),
            },
        }
    }
}

impl<'a, T> IfBlock<'a, T> {
    pub fn inner_recursive(&'a self) -> NodeIter<'a, T> {
        NodeIter::new(&self.inner)
    }

    pub fn else_recursive(&'a self) -> Option<NodeIter<'a, T>> {
        self.else_block
            .as_ref()
            .map(|else_block| NodeIter::new(else_block))
    }
}

impl<'a> DecorousAst<'a> {
    /// Creates a recursive iterator over the nodes of the template.
    pub fn iter_nodes(&'a self) -> NodeIter<'a, Location> {
        NodeIter::new(&self.nodes)
    }
}

pub fn traverse_with<'a, T, F, G>(nodes: &'a [Node<'a, T>], predicate: &mut F, body_func: &mut G)
where
    F: FnMut(&Element<'a, T>) -> bool,
    G: FnMut(&Node<'a, T>),
{
    for node in nodes {
        body_func(node);
        if let NodeType::Element(elem) = &node.node_type {
            if !predicate(elem) {
                continue;
            }
            traverse_with(&elem.children, predicate, body_func);
        }
    }
}

pub fn traverse_mut<'a, T, F>(nodes: &mut [Node<'a, T>], f: &mut F)
where
    F: FnMut(&mut Node<'a, T>),
{
    for node in nodes {
        f(node);
        if let NodeType::Element(elem) = &mut node.node_type {
            traverse_mut(&mut elem.children, f);
        }
    }
}

#[derive(Debug)]
pub struct NodeIter<'a, T> {
    stack: Vec<&'a Node<'a, T>>,
}

impl<'a, T> NodeIter<'a, T> {
    pub fn new(nodes: &'a [Node<'a, T>]) -> NodeIter<'a, T> {
        let mut stack = Vec::with_capacity(nodes.len());
        stack.extend(nodes.iter().rev());
        Self { stack }
    }
}

impl<'a, T> Iterator for NodeIter<'a, T> {
    type Item = &'a Node<'a, T>;

    fn next(&mut self) -> Option<Self::Item> {
        self.stack.pop().map(|node| {
            if let NodeType::Element(elem) = &node.node_type {
                self.stack.extend(elem.children.iter().rev());
            }

            node
        })
    }
}

impl<'a, T> fmt::Display for Node<'a, T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match &self.node_type {
            NodeType::Text(t) => write!(f, "{t}"),
            NodeType::Comment(Comment(c)) => write!(f, "<!--{c}-->"),
            NodeType::Element(elem) => write!(f, "{elem}"),
            NodeType::Mustache(js) => write!(f, "{{{js}}}"),
            NodeType::SpecialBlock(block) => write!(f, "{block}"),
        }
    }
}

impl<'a, T> fmt::Display for Element<'a, T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "<{}{}{}>{}</{0}>",
            self.tag,
            (!self.children.is_empty())
                .then_some(" ")
                .unwrap_or_default(),
            self.attrs.iter().join(" "),
            self.children
                .iter()
                .map(|elem| format!("  {elem}"))
                .join("")
        )
    }
}

impl<'a> fmt::Display for Attribute<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Attribute::KeyValue(key, Some(val)) => write!(f, "{key}={val}"),
            Attribute::KeyValue(key, None) => write!(f, "{key}"),
            Attribute::EventHandler(event_handler) => write!(f, "{event_handler}"),
            Attribute::Binding(binding) => write!(f, "bind:{binding}"),
        }
    }
}

impl<'a> fmt::Display for AttributeValue<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            AttributeValue::Literal(literal) => write!(f, "\"{literal}\""),
            AttributeValue::JavaScript(js) => write!(f, "{{{js}}}"),
        }
    }
}

impl<'a> fmt::Display for EventHandler<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "on:{}={{{}}}", self.event, self.expr)
    }
}

impl<'a, T> fmt::Display for SpecialBlock<'a, T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            SpecialBlock::If(if_block) => write!(f, "{if_block}"),
            SpecialBlock::For(for_block) => write!(f, "{for_block}"),
            SpecialBlock::Use(use_block) => write!(f, "{use_block}"),
        }
    }
}

impl<'a, T> fmt::Display for IfBlock<'a, T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{{#if {}}}\n{}\n{{/if}}",
            self.expr,
            self.inner.iter().map(|elem| format!("  {elem}")).join(""),
        )
    }
}

impl<'a, T> fmt::Display for ForBlock<'a, T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{{#for {} in {}}}\n{}\n{{/for}}",
            self.index.map_or_else(
                || self.binding.to_owned(),
                |idx| format!("{idx}, {}", self.binding)
            ),
            self.expr,
            self.inner.iter().map(|elem| format!("  {elem}")).join(""),
        )
    }
}

impl<'a> fmt::Display for UseBlock<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{{#use \"{}\"}}", self.path.display())
    }
}
