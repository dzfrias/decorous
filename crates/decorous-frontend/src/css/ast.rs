use std::{borrow::Cow, fmt};

use itertools::Itertools;
use rslint_parser::SyntaxNode;

#[derive(Debug)]
pub struct Css<'a> {
    rules: Vec<Rule<'a>>,
}

#[derive(Debug, PartialEq)]
pub enum Rule<'a> {
    At(AtRule<'a>),
    Regular(RegularRule<'a>),
}

#[derive(Debug, PartialEq)]
pub struct RegularRule<'a> {
    selector: Selector<'a>,
    declarations: Vec<Declaration<'a>>,
}

#[derive(Debug, PartialEq)]
pub struct AtRule<'a> {
    name: &'a str,
    additional: &'a str,
    contents: Option<Vec<Rule<'a>>>,
}

#[derive(Debug, PartialEq)]
pub struct Selector<'a> {
    parts: Vec<SelectorPart<'a>>,
}

#[derive(Debug, PartialEq)]
pub struct SelectorPart<'a> {
    text: Option<Cow<'a, str>>,
    pseudoes: Vec<Pseudo<'a>>,
}

#[derive(Debug, PartialEq)]
pub enum Pseudo<'a> {
    Element(&'a str),
    Class {
        name: &'a str,
        value: Option<&'a str>,
    },
}

#[derive(Debug, PartialEq)]
pub struct Declaration<'a> {
    name: &'a str,
    values: Vec<Value<'a>>,
}

#[derive(Debug, PartialEq)]
pub enum Value<'a> {
    Mustache(SyntaxNode),
    Css(&'a str),
}

impl<'a> Declaration<'a> {
    pub fn new(name: &'a str, values: Vec<Value<'a>>) -> Self {
        Self { name, values }
    }

    pub fn name(&self) -> &str {
        self.name
    }

    pub fn values(&self) -> &[Value<'_>] {
        self.values.as_ref()
    }

    pub fn values_mut(&mut self) -> &mut Vec<Value<'a>> {
        &mut self.values
    }

    pub fn name_mut(&mut self) -> &mut &'a str {
        &mut self.name
    }
}

impl<'a> SelectorPart<'a> {
    pub fn new(text: Option<Cow<'a, str>>, pseudoes: Vec<Pseudo<'a>>) -> Self {
        Self { text, pseudoes }
    }

    pub fn text(&self) -> Option<&Cow<'a, str>> {
        self.text.as_ref()
    }

    pub fn pseudoes(&self) -> &[Pseudo<'_>] {
        self.pseudoes.as_ref()
    }

    pub fn text_mut(&mut self) -> &mut Option<Cow<'a, str>> {
        &mut self.text
    }

    pub fn pseudoes_mut(&mut self) -> &mut Vec<Pseudo<'a>> {
        &mut self.pseudoes
    }
}

impl<'a> Selector<'a> {
    pub fn new(parts: Vec<SelectorPart<'a>>) -> Self {
        Self { parts }
    }

    pub fn parts_mut(&mut self) -> &mut Vec<SelectorPart<'a>> {
        &mut self.parts
    }

    pub fn parts(&self) -> &[SelectorPart<'_>] {
        self.parts.as_ref()
    }
}

impl<'a> RegularRule<'a> {
    pub fn new(selector: Selector<'a>, declarations: Vec<Declaration<'a>>) -> Self {
        Self {
            selector,
            declarations,
        }
    }

    pub fn selector(&self) -> &Selector<'a> {
        &self.selector
    }

    pub fn declarations(&self) -> &[Declaration<'_>] {
        self.declarations.as_ref()
    }

    pub fn selector_mut(&mut self) -> &mut Selector<'a> {
        &mut self.selector
    }

    pub fn declarations_mut(&mut self) -> &mut Vec<Declaration<'a>> {
        &mut self.declarations
    }
}

impl<'a> Css<'a> {
    pub fn new(rules: Vec<Rule<'a>>) -> Self {
        Self { rules }
    }

    pub fn rules(&self) -> &[Rule<'a>] {
        &self.rules
    }

    pub fn rules_mut(&mut self) -> &mut Vec<Rule<'a>> {
        &mut self.rules
    }
}

impl<'a> AtRule<'a> {
    pub fn new(name: &'a str, additional: &'a str, contents: Option<Vec<Rule<'a>>>) -> Self {
        Self {
            name,
            additional,
            contents,
        }
    }

    pub fn name(&self) -> &str {
        self.name
    }

    pub fn additional(&self) -> &str {
        self.additional
    }

    pub fn contents(&self) -> Option<&[Rule<'a>]> {
        self.contents.as_deref()
    }

    pub fn contents_mut(&mut self) -> &mut Option<Vec<Rule<'a>>> {
        &mut self.contents
    }

    pub fn additional_mut(&mut self) -> &mut &'a str {
        &mut self.additional
    }

    pub fn name_mut(&mut self) -> &mut &'a str {
        &mut self.name
    }
}

// ---Display impls---

impl<'a> fmt::Display for SelectorPart<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{}{}",
            self.text().unwrap_or(&Cow::default()),
            self.pseudoes().iter().join("")
        )
    }
}

impl<'a> fmt::Display for Value<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Mustache(s) => write!(f, "{s}"),
            Self::Css(css) => write!(f, "{css}"),
        }
    }
}

impl<'a> fmt::Display for Pseudo<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Element(elem) => write!(f, "::{elem}"),
            Self::Class {
                name,
                value: Some(value),
            } => write!(f, ":{name}({value})"),
            Self::Class { name, value: None } => {
                write!(f, ":{name}")
            }
        }
    }
}

impl<'a> fmt::Display for Declaration<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}: {};", self.name, self.values().iter().join(" "))
    }
}

impl<'a> fmt::Display for AtRule<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        if let Some(contents) = self.contents() {
            write!(
                f,
                "@{} {} {{ {} }}",
                self.name(),
                self.additional(),
                contents.iter().join(" ")
            )
        } else {
            write!(f, "@{} {};", self.name(), self.additional())
        }
    }
}

impl<'a> fmt::Display for RegularRule<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{} {{ {} }}",
            self.selector(),
            self.declarations().iter().join("; ")
        )
    }
}

impl<'a> fmt::Display for Rule<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::At(at_rule) => write!(f, "{at_rule}"),
            Self::Regular(regular) => write!(f, "{regular}"),
        }
    }
}

impl<'a> fmt::Display for Selector<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.parts().iter().join(" "))
    }
}

impl<'a> fmt::Display for Css<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.rules().iter().join("\n"))
    }
}
