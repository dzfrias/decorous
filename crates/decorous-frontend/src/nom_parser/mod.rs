#![allow(dead_code, unused_imports)]

mod errors;

use nom::{
    branch::alt,
    bytes::complete::{escaped, tag, take, take_until, take_while, take_while1},
    character::complete::{
        alpha1, alphanumeric1, anychar, char, multispace0, multispace1, none_of, one_of,
    },
    combinator::{cut, eof, map, opt, recognize, value},
    multi::{many0, many0_count, separated_list0},
    sequence::{delimited, pair, preceded, terminated},
    IResult,
};
use nom_locate::{position, LocatedSpan};
use rslint_parser::{parse_module, SyntaxNode};

use self::errors::{ParseError, ParseErrorType, Report};
use crate::ast::{Attribute, AttributeValue, Element, EventHandler, Location, Node, NodeType};

type Result<'a, Output> = IResult<NomSpan<'a>, Output, Report<NomSpan<'a>>>;
type NomSpan<'a> = LocatedSpan<&'a str>;

#[derive(Debug)]
pub struct NewLocation {
    offset: usize,
    length: usize,
    line: u32,
    column: usize,
}

impl NewLocation {
    pub fn from_spans(span1: NomSpan, span2: NomSpan) -> Self {
        Self {
            offset: span1.location_offset(),
            length: span2.location_offset() - span1.location_offset(),
            line: span1.location_line(),
            column: span1.get_column(),
        }
    }

    pub fn offset(&self) -> usize {
        self.offset
    }

    pub fn length(&self) -> usize {
        self.length
    }

    pub fn line(&self) -> u32 {
        self.line
    }

    pub fn column(&self) -> usize {
        self.column
    }
}

// pub fn parse<'a>(input: NomSpan<'a>) -> Result<Vec<Node<'a, NewLocation>>> {
//     let (input, script) = opt(ws(script))(input)?;
//     let (input, nodes) = nodes(input)?;
//     // TODO: Turn into decorous ast
// }

fn failure_case<'i, Free, Parsed, P, F>(
    parser: P,
    err_constructor: F,
) -> impl Fn(NomSpan<'i>) -> Result<'i, Free>
where
    P: Fn(NomSpan<'i>) -> Result<'i, Parsed>,
    F: Fn(Parsed) -> ParseErrorType,
{
    use nom::Err::{Error, Failure, Incomplete};
    move |i: NomSpan<'i>| match parser(i) {
        Ok((_, parsed)) => {
            let report = Report::from(ParseError::new(i, err_constructor(parsed)));
            Err(Failure(report))
        }
        Err(Failure(e)) => Err(Failure(e)),
        Err(Error(e)) => Err(Error(e)),
        Err(Incomplete(need)) => Err(Incomplete(need)),
    }
}

fn bad_char(input: NomSpan) -> Result<char> {
    alt((value('\0', eof), anychar))(input)
}

fn parse_str(i: NomSpan) -> Result<NomSpan> {
    escaped(none_of("\""), '\\', one_of("\"n\\"))(i)
}

fn any_word(input: NomSpan) -> Result<NomSpan> {
    recognize(take_while(|c: char| c.is_whitespace()))(input)
}

fn script(input: NomSpan) -> Result<SyntaxNode> {
    let (input, _) = tag("---js")(input)?;
    let (input, body) = take_until("---")(input)?;
    let (input, _) = take(3usize)(input)?;
    // TODO: Errors
    let parsed = parse_module(&body, 0);
    Ok((input, parsed.syntax()))
}

fn nodes(input: NomSpan) -> Result<Vec<Node<'_, NewLocation>>> {
    many0(node)(input)
}

fn node(input: NomSpan) -> Result<Node<'_, NewLocation>> {
    let (input, pos) = position(input)?;
    let (input, node) = alt((
        map(element, NodeType::Element),
        map(comment, |c| NodeType::Comment(&c)),
        map(mustache, NodeType::Mustache),
        map(
            take_while1(|c| c != '/' && c != '#' && c != '{'),
            |text: NomSpan| NodeType::Text(&text),
        ),
    ))(input)?;
    let (input, end_pos) = position(input)?;
    let location = NewLocation::from_spans(pos, end_pos);

    Ok((input, Node::new(node, location)))
}

fn element(input: NomSpan) -> Result<Element<'_, NewLocation>> {
    let (input, _) = tag("#")(input)?;
    let (input, tag_name) = alphanumeric1(input)?;
    let (input, attrs) = terminated(
        opt(delimited(
            ws_trailing(tag("[")),
            separated_list0(multispace1, attribute),
            alt((
                ws_leading(tag("]")),
                failure_case(bad_char, |_| ParseErrorType::ExpectedCharacter(']')),
            )),
        )),
        multispace0,
    )(input)?;
    let (input, children) = nodes(input)?;
    let (input, _) = preceded(
        char('/'),
        alt((
            tag(*tag_name.fragment()),
            failure_case(identifier, |tag| {
                ParseErrorType::InvalidClosingTag(tag.to_string())
            }),
        )),
    )(input)?;
    Ok((
        input,
        Element::new(&tag_name, attrs.unwrap_or_default(), children),
    ))
}

fn attribute(input: NomSpan) -> Result<Attribute<'_>> {
    let (input, event_handler) = map(opt(char('@')), |evt| evt.is_some())(input)?;
    let (input, attr) = identifier(input)?;
    if event_handler {
        let (input, _) = alt((
            ws(char('=')),
            failure_case(bad_char, |_| ParseErrorType::ExpectedCharacter('=')),
        ))(input)?;
        return map(mustache, |node| {
            Attribute::EventHandler(EventHandler::new(&attr, node))
        })(input);
    }

    let (input, eq_char) = opt(ws(char('=')))(input)?;
    if eq_char.is_none() {
        return Ok((input, Attribute::KeyValue(&attr, None)));
    }
    let (input, attribute) = alt((
        map(mustache, |node| {
            Attribute::KeyValue(&attr, Some(AttributeValue::JavaScript(node)))
        }),
        map(
            preceded(
                char('\"'),
                alt((
                    terminated(parse_str, char('\"')),
                    failure_case(bad_char, |_| ParseErrorType::ExpectedCharacter('"')),
                )),
            ),
            |t| Attribute::KeyValue(&attr, Some(AttributeValue::Literal(&t))),
        ),
    ))(input)?;

    Ok((input, attribute))
}

fn comment(input: NomSpan) -> Result<NomSpan> {
    preceded(tag("//"), take_while(|c| c != '\n'))(input)
}

fn mustache(input: NomSpan) -> Result<SyntaxNode> {
    // FIX: Allow for curly braces in JavaScript
    let (input, js_text) = delimited(tag("{"), take_while(|c| c != '}'), tag("}"))(input)?;
    Ok((input, parse_module(&js_text, 0).syntax()))
}

fn ws<'a, F: 'a, O>(inner: F) -> impl FnMut(NomSpan<'a>) -> Result<O>
where
    F: Fn(NomSpan<'a>) -> Result<O>,
{
    delimited(multispace0, inner, multispace0)
}

fn ws_leading<'a, F: 'a, O>(inner: F) -> impl FnMut(NomSpan<'a>) -> Result<O>
where
    F: Fn(NomSpan<'a>) -> Result<O>,
{
    preceded(multispace0, inner)
}

fn ws_trailing<'a, F: 'a, O>(inner: F) -> impl FnMut(NomSpan<'a>) -> Result<O>
where
    F: Fn(NomSpan<'a>) -> Result<O>,
{
    terminated(inner, multispace0)
}

fn identifier(input: NomSpan) -> Result<NomSpan> {
    recognize(pair(
        alt((alpha1, tag("_"))),
        many0_count(alt((alphanumeric1, tag("_")))),
    ))(input)
}

#[cfg(test)]
mod tests {
    use super::*;

    macro_rules! nom_test_all_insta {
        ($func:ident, $inputs:expr) => {
            for input in $inputs {
                insta::assert_debug_snapshot!($func(input.into()));
            }
        };
    }

    #[test]
    fn can_parse_attribute() {
        nom_test_all_insta!(
            attribute,
            [
                "hello=\"hello world\"",
                "hello",
                "hello   =   \"hello world\"",
                "hello   =   \"hello world\"",
                "hello=\"你好\"",
                "hello={ctx[1]}",
                "@click={() => x += 1}",
                "@click   =  \"wrong\"",
                "@click",
            ]
        );
    }

    #[test]
    fn can_parse_elements() {
        nom_test_all_insta!(
            element,
            [
                "#div /div",
                "#div[hello=\"world\"]/div",
                "#div[   hello=\"world\"    ] /div",
                "#div[   hello=\"world    ] /div",
                "#div[]/div",
                "#div#li/li/div",
                "#div #ul    /ul/div",
                "#div text {mustache} hello #div/div /div",
                "#div[hello=\"world\"/div",
                "#div[hello=\"world\"]/notdiv",
                "#div]/div",
                "#div"
            ]
        );
    }

    #[test]
    fn can_parse_multiple_elements() {
        nom_test_all_insta!(
            nodes,
            [
                "#div/div#div hello /div hello",
                "",
                "hello",
                "/div",
                "#div/div// hello!"
            ]
        );
    }

    #[test]
    fn can_parse_scripts() {
        nom_test_all_insta!(script, ["---js console.log(\"hello\")---"])
    }
}
