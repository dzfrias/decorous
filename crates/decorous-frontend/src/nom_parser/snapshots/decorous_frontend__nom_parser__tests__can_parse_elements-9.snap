---
source: crates/decorous-frontend/src/nom_parser/mod.rs
expression: element(input.into())
---
Err(
    Failure(
        Report {
            errors: [
                ParseError {
                    fragment: LocatedSpan {
                        offset: 18,
                        line: 1,
                        fragment: "/div",
                        extra: (),
                    },
                    err_type: ExpectedCharacter(
                        ']',
                    ),
                },
            ],
        },
    ),
)
